### PR TITLE
[SimpleFSDP] Add support for SimpleFSDP DCP

### PR DESCRIPTION
**Pull request by ruisizhang123**
As titled, this pr adds support for SimpleFSDP's DCP composability. The code is based on previous implementations from @fmassa @awgu @yf225 

The following losses match perfectly. The checkpoints are loaded from step-110.

(1) [dp:4] --> [dp:4]

Screenshot 2025-06-08 at 5 39 50 PM

(2) [dp:2, tp:2] --> [dp:2, tp:2] & [dp:2, pp:2]

Screenshot 2025-06-08 at 5 46 29 PM

### CODE REVIEW

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py [hunk 1]
```diff
@@ -124,6 +124,25 @@ def _distribute_dtensor(
...
+            orig_param = self._parameters[param_name]
+            return parametrization(orig_param)
+
+        return param
+
+    param_name_to_property = {param_name: property(param_fn(param_name))}
+    param_cls = type(
+        f"FSDP{module.__class__.__name__}",
+        (module.__class__,),
+        param_name_to_property,
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/simple_fsdp.py:142]: I feel this won't work, as it's a dict containing only one parameter, when the module can have many parameters. The reason is that you are doing it inside the for loop over parameters. Hmm but I'm not sure why the code doesn't fail .. I need to run some examples to understand.

I was thinking that we need to do it outside `for p_name, p in params_dict.items():`. You can take another look at Andrew's code or Will's PR https://github.com/pytorch/pytorch/pull/140360.

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py [hunk 2]
```diff
@@ -263,7 +282,7 @@ def data_parallel(
                         distribute_tensor_func(p, device_mesh, param_sharding)
                     ),
                 )
-                nn.utils.parametrize.register_parametrization(
+                _register_parametrization(
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/simple_fsdp.py]: Comment around that we are not using `nn.utils.parametrize.register_parametrization` because it's not compatible with checkpointing. We can even leave the old code here, commented, as reference.

**Review by tianyu-l — COMMENTED** [review]
Please give credit to Francisco and Andrew in the PR summary.

Please include comparisons of loss curves with and without loading from a checkpoint, including at least the following sets:
1. DP4 -> DP4
2. DP2 -> DP2 TP2
4. DP2 -> DP2 PP2

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py (hunk 1 above)
Comment (ruisizhang123) [reply torchtitan/experiments/simple_fsdp/simple_fsdp.py:142]: hmmmm it turns out `params_dict` only has one key name `weight`. That's why this works.

I should move this outside the for loop.

Comment (tianyu-l) [reply torchtitan/experiments/simple_fsdp/simple_fsdp.py:142]: oh good point, I just realized in the llama3 model definition, there is no module with even a `bias` parameter.

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py [hunk 3]
```diff
@@ -124,6 +124,27 @@ def _distribute_dtensor(
...
+
+    param_name_to_property = {
+        param_name: property(param_fn(param_name)) for param_name in param_names
+    }
+    param_cls = type(
+        f"FSDP{module.__class__.__name__}",
+        (module.__class__,),
+        param_name_to_property,
+    )
+    module.__class__ = param_cls
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/simple_fsdp.py]: I think we can reuse the idea of partial function, to avoid defining a 3-layer recursive function.
Let's also rename `param_cls` to `module_cls`.
```suggestion
    def param_fn(self, param_name: str):
        orig_param = self._parameters[param_name]
        return parametrization(orig_param)

    param_name_to_property = {
        param_name: property(partial(param_fn, param_name=param_name))
        for param_name in param_names
    }
    module_cls = type(
        f"FSDP{module.__class__.__name__}",
        (module.__class__,),
        param_name_to_property,
    )
    module.__class__ = module_cls
```

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py [hunk 4]
```diff
@@ -124,6 +124,27 @@ def _distribute_dtensor(
     )
 
 
+def _register_parametrization(
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/simple_fsdp.py:127]: add a comment like:
The reason this function works with state_dict without incurring parametrization calls is because state_dict accesses parameters directly from self._parameters, not from getters.
https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/module.py#L2141

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py (hunk 3 above)
Comment (ruisizhang123) [reply torchtitan/experiments/simple_fsdp/simple_fsdp.py]: Seems like `partial` function cannot be traced here: https://github.com/pytorch/pytorch/issues/43520. 

I'm updating with `lambda` funcs.

File: torchtitan/experiments/simple_fsdp/simple_fsdp.py [hunk 5]
```diff
@@ -124,6 +124,28 @@ def _distribute_dtensor(
...
+    it works with state_dict without incurring parametrization calls because
+    state_dict accesses parameters directly from self._parameters, not from getters
+    https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/module.py#L2141
+    """
+    param_name_to_property = {
+        param_name: property(
+            lambda self, name=param_name: parametrization(self._parameters[name])
+        )
+        for param_name in param_names
+    }
```
Comment (tianyu-l) [inline torchtitan/experiments/simple_fsdp/simple_fsdp.py]: the `name` parameter seems not necessary, so we can make it as simple as
```suggestion
    param_name_to_property = {
        param_name: property(lambda self: parametrization(self._parameters[param_name]))
        for param_name in param_names
    }
```

Comment (ruisizhang123) [reply torchtitan/experiments/simple_fsdp/simple_fsdp.py]: sounds good

**Comment by fegin** [comment]
uh, this may not always work: https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/state_dict.py#L223

**Comment by ruisizhang123** [comment]
> uh, this may not always work: https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/state_dict.py#L223

If I understand correctly, this code is to ensure the original FQN unchanged. Without this function, when reparametrize  the model, for example, the weight dict key would be updated from `model.tok_embeddings.weight` to `model.tok_embeddings.parametrizations.weight.original`. 

After applying it, the weight dict keeps the original `model.tok_embeddings.weight`, while applying reparametrization. When reloading  the checkpoints with `_get_fqns`, applying `getattr` should be good. Because we are still accessing `model.tok_embeddings.weight`. Maybe I should update comments to avoid confusion 🤔️

cc @tianyu-l

**Comment by fegin** [comment]
@ruisizhang123 I think it will work but the problem is that whether there will be extra allgather as we don't want communication during checkpointing.

**Comment by fegin** [comment]
I reviewed the code again, the logic is a check so we won't change anything unless the FQNs have mismatch. So the returned state_dict will have the correct FQNs and values (DTensor). The potential issue is that the whether the check will cause extra unnecessary allgather, my read is yes.

Overall, the implementation is correct and can support DCP, but we will have to verify whether there are unnecessary allgather during save and load checkpoints.

cc., @ruisizhang123 @tianyu-l

**Comment by ruisizhang123** [comment]
> I reviewed the code again, the logic is a check so we won't change anything unless the FQNs have mismatch. So the returned state_dict will have the correct FQNs and values (DTensor). The potential issue is that the whether the check will cause extra unnecessary allgather, my read is yes.
> 
> Overall, the implementation is correct and can support DCP, but we will have to verify whether there are unnecessary allgather during save and load checkpoints.
> 
> cc., @ruisizhang123 @tianyu-l

Good point. Seems there are additional AG when doing DCP. Let's see how to fix it.
